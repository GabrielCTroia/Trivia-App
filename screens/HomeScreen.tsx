import React from 'react';
import {
  StyleSheet,
  View,
  Text,
} from 'react-native';

import R from 'ramda';
import * as Colors from '../styles/Colors';
import { StyledButton } from '../components/buttons/StyledButton';
import { NavigationScreenProp } from 'react-navigation';
import { getQuestionsByCategory, Question } from '../Api/Questions';
import * as Typography from '../styles/Typography';
import { screenLayout, screenVerticalPadding } from '../styles/Layout';

export interface HomeScreenProps {
  navigation: NavigationScreenProp<any, any>;
}

interface State {
  questionsByCategories: { [k: string]: Question[] },
}

export class HomeScreen extends React.Component<HomeScreenProps, State> {
  static navigationOptions = {
    header: null,
  };

  constructor(props: HomeScreenProps) {
    super(props);

    this.state = {
      questionsByCategories: {},
    };
  }

  // async componentDidMount() {
  //   await this.fetchData();

  //   this.navigateToQuizScreen(R.keys(this.state.questionsByCategories)[0].toString())
  // }

  private async fetchData() {
    this.setState({
      questionsByCategories: await getQuestionsByCategory(),
    });
  }

  private navigateToQuizScreen(category: string) {
    const { navigate } = this.props.navigation;

    navigate('Quiz', {
      category,
      questions: this.state.questionsByCategories[category],
    });
  }

  private showCategories() {
    const categories = R.keys(this.state.questionsByCategories);

    if (categories.length === 0) {
      return null;
    }

    return (
      <View>
        {categories.map((c, i) => (
          <StyledButton
            key={i}
            title={c.toString()}
            onPress={() => this.navigateToQuizScreen(c.toString())}
          />
        ))}
      </View>
    );
  }

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.main}>
          <View>
            <Text style={styles.header}>Welcome to</Text>
            <Text style={styles.subheader}>Trivia Challenge</Text>
          </View>
          <Text style={styles.baseText}>
            You will be presented with 10 True or False questions.
          </Text>
          <Text style={[styles.baseText, Typography.centeredText]}>
            Can you score 100%?
          </Text>
        </View>

        <View style={styles.quizIntro}>
          <StyledButton
            title="Begin"
            onPress={() => { this.fetchData() }}
          />
          {this.showCategories()}
        </View>
      </View>
    );
  }

}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundColor,
  },
  main: {
    ...screenLayout,
    paddingBottom: screenVerticalPadding,

    justifyContent: 'space-between',

    flex: .85,
  },
  header: {
    ...Typography.headerText,
    color: Colors.baseTextInvertedColor,
  },
  subheader: {
    ...Typography.subheaderText,
    color: Colors.subheaderTextColor,
  },
  baseText: {
    ...Typography.baseText,
    color: Colors.baseTextInvertedColor,
    fontSize: 24,
  },

  quizIntro: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,

    backgroundColor: Colors.foregroundColor,
    padding: 20,
    borderTopLeftRadius: 45,
    borderTopRightRadius: 45,

    alignContent: "center",
    justifyContent: 'space-evenly',

    shadowColor: '#D0BD0F',
    shadowOffset: { width: 0, height: 4 },
    shadowRadius: 10,
    shadowOpacity: 1,
  }
});
