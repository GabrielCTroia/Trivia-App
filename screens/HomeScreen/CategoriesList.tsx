import React, { FunctionComponent } from 'react';
import { View, Text, FlatList, StyleSheet } from 'react-native';
import { TouchableListItem } from '../../components/List/TouchableListItem';
import * as Typography from '../../styles/Typography';
import * as Colors from '../../styles/Colors';

export type CategoriesListProps = {
  categories: string[];
  onPress: (category: string) => void;
}

export const CategoriesList: FunctionComponent<CategoriesListProps> = (props) => (
  <View>
    <Text style={styles.categoriesListHeader}>Choose Category</Text>
    <FlatList
      data={props.categories}
      keyExtractor={(_, i) => i.toString()}
      renderItem={({ item }) => (
        <TouchableListItem
          content={item}
          onPress={() => { props.onPress(item) }}
        />
      )}
    />
  </View>
);

const styles = StyleSheet.create({
  categoriesListHeader: {
    ...Typography .baseText,
    ...Typography.boldText,
    fontSize: 19,
    paddingBottom: 20,
    color: Colors.baseTextColor,
    fontWeight: 'bold',
  }
});
