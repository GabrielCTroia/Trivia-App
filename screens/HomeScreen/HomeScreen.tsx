import React from 'react';
import { StyleSheet, View, Text, StatusBar } from 'react-native';
import { StyledButton, ButtonTypes } from '../../components/buttons/StyledButton';
import { NavigationScreenProp } from 'react-navigation';
import { getQuestionsByCategory, Question } from '../../Api/Questions';
import { CategoriesList } from './CategoriesList';
import * as Colors from '../../styles/Colors';
import * as Typography from '../../styles/Typography';
import * as Layout from '../../styles/Layout';


export interface HomeScreenProps {
  navigation: NavigationScreenProp<any, any>;
}

interface State {
  questionsByCategories: { [k: string]: Question[] },
  fetching: boolean;
  fetchError: boolean;
}

export class HomeScreen extends React.Component<HomeScreenProps, State> {
  static navigationOptions = {
    header: null,
  };

  constructor(props: HomeScreenProps) {
    super(props);

    this.state = {
      questionsByCategories: {},
      fetching: false,
      fetchError: false,
    };
  }

  private async fetchData() {
    await this.setState({ fetching: true });

    getQuestionsByCategory()
      .then((questionsByCategories) => {
        this.setState({
          fetching: false,
          questionsByCategories,
          fetchError: false,
        });
      })
      .catch(() => {
        this.setState({
          fetching: false,
          fetchError: true,
        });
      });
  }

  private navigateToQuizScreen(category: string) {
    const { navigate } = this.props.navigation;

    navigate('Quiz', {
      category,
      questions: this.state.questionsByCategories[category],
    });
  }

  private showBeginButton() {
    const categories = Object.keys(this.state.questionsByCategories);

    if (categories.length > 0) {
      return null;
    }

    return (
      <StyledButton
        title="Begin"
        buttonType={ButtonTypes.warning}
        buttonFit="full"
        onPress={() => { this.state.fetching || this.fetchData() }}
      />
    );
  }

  private showCategories() {
    const categories = Object.keys(this.state.questionsByCategories);

    if (categories.length === 0) {
      return null;
    }

    return (
      <CategoriesList
        categories={categories.map((c) => c.toString())}
        onPress={(item) => this.navigateToQuizScreen(item)}
      />
    );
  }

  private showFetchError() {
    if (!this.state.fetchError || this.state.fetching) {
      return null;
    }

    return (
      <View style={styles.error}>
        <Text style={styles.errorText}>
          There was a Network Error. Please try again!
        </Text>
      </View>
    )
  }

  render() {
    return (
      <View style={styles.container}>
        <StatusBar barStyle="light-content" />
        <View style={styles.main}>
          <View>
            <Text style={styles.header}>Welcome to</Text>
            <Text style={styles.subheader}>Trivia Challenge</Text>
          </View>

          <View style={styles.tagLines}>
            <Text style={styles.baseText}>
              You will be presented with 10 True or False questions.
              </Text>
            <Text style={[styles.baseText, Typography.boldText]}>
              Can you score 100%?
            </Text>
          </View>
        </View>

        <View style={styles.quizIntro}>
          {this.showBeginButton()}
          {this.showCategories()}
          {this.showFetchError()}
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundColor,
  },
  main: {
    ...Layout.screenLayoutWithoutHeader,
    flex: .8,
    height: Layout.screenHeight - 120,
    paddingBottom: Layout.screenVerticalPadding,
  },
  tagLines: {
    paddingTop: 100,
  },
  header: {
    ...Typography.headerText,
    color: Colors.baseTextColorInverted,
    fontSize: Typography.largestFontSize,
  },
  subheader: {
    ...Typography.subheaderText,
    color: Colors.subheaderTextColorInverted,
    fontSize: Typography.extraLargeFontSize,
  },
  baseText: {
    ...Typography.baseText,
    color: Colors.baseTextColorInverted,
    fontSize: 24,
  },
  quizIntro: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,

    paddingTop: Layout.insetPadding * 2,
    paddingBottom: Layout.insetPadding * 2,
    paddingLeft: Layout.insetPadding,
    paddingRight: Layout.insetPadding,

    backgroundColor: Colors.foregroundColor,
    borderTopLeftRadius: Layout.screenBorderRadius,
    borderTopRightRadius: Layout.screenBorderRadius,

    alignContent: "center",
    justifyContent: 'space-evenly',
  },
  error: {
    paddingTop: 20,
  },
  errorText: {
    ...Typography.errorText,
    textAlign: 'center',
  }
});
