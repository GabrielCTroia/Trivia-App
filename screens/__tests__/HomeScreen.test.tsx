import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { HomeScreen } from '../HomeScreen/HomeScreen';

it('renders correctly', () => {
  const tree = renderer.create(<HomeScreen />).toJSON();

  expect(tree).toMatchSnapshot();
});
