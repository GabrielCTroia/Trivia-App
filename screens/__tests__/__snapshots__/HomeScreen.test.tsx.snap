// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "backgroundColor": "#cccccc",
      "flex": 1,
      "justifyContent": "space-between",
      "paddingTop": 20,
    }
  }
>
  <View>
    <Text
      style={
        Array [
          Object {},
          Object {
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "center",
          },
        ]
      }
    >
      Welcome to Trivia Challenge!
    </Text>
  </View>
  <View>
    <Text
      style={
        Array [
          Object {},
          Array [
            Object {
              "fontSize": 20,
            },
            Object {
              "textAlign": "center",
            },
          ],
        ]
      }
    >
      You will be presented with 10 True or False questions.
    </Text>
  </View>
  <View>
    <Text
      style={
        Array [
          Object {},
          Array [
            Object {
              "fontSize": 20,
            },
            Object {
              "textAlign": "center",
            },
          ],
        ]
      }
    >
      Can you score 100%?
    </Text>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignContent": "center",
            "padding": 10,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "textAlign": "center",
            "textTransform": "uppercase",
          }
        }
      >
        Begin
      </Text>
    </View>
  </View>
</View>
`;
